test/test_buffer_protocol.py

```diff
@@ -0,0 +1,184 @@
+import torch.testing._internal.common_utils as common
+from torch.testing._internal.common_device_type import (
+    instantiate_device_type_tests,
+    dtypes, onlyCPU
+)
+from torch.testing import all_types
+
+import torch
+import array
+
+TORCH_TO_ARRAYTYPE = {
+    torch.int8    : 'b',
+    torch.uint8   : 'B',
+    torch.int16   : 'h',
+    torch.int32   : 'i',
+    torch.int64   : 'q',
+    torch.float32 : 'f',
+    torch.float64 : 'd',
```

Comment:
Maybe I have a bad memory but: why no `complex128`, `complex256`? Worth a comment if it's left out on purpose.

Reply:
Because [the python array module](https://docs.python.org/3/library/array.html) doesn't have typecodes for those.

Reply:
Ah, but this is the buffer protocol - which does have them: https://docs.python.org/3/c-api/buffer.html#complex-arrays. The buffer protocol docs are terrible, and don't give you these type codes. Maybe easiest to look at the source code for it, or for how NumPy processes the buffer protocol.

Reply:
Huh. Maybe I'm missing something, but I can't find where it describes the complex types (as far as I understood it, it describes how to deal with arrays with strides and suboffsets).

Anyways, this table is used for converting PyTorch dtypes into Python Array module's typecodes (and, apparently, it doesn't support complex types). But, maybe, instead of the array module, I could be using NumPy (since it does implement the buffer protocol) as [suggested by @mruberry previously](https://github.com/pytorch/pytorch/pull/59077#discussion_r646091867).